design-docs/operationBased2.md

```diff
@@ -0,0 +1,269 @@
+<td><pre lang="kotlin">
+public data class Data(public val animal: Animal?)
+data class Animal(
+  val lionFragment: LionFragment?,
+  val catFragment: CatFragment?,
+)
+</pre></td>
+</tr>
+</table>
+
+
+## Example 3
+
+<table>
+<tr><th>GraphQL</th><th>Kotlin</th></tr>
+<tr>
+<td><pre lang="graphql">
+query TestOperation {
+  animal {
+    # Field type = Inline fragment type
+    ... on Animal {
+      species
+    }
+  }
+}
+</pre></td>
+<td><pre lang="kotlin">
+public data class Data(public val animal: Animal?)
+data class Animal(val onAnimal: OnAnimal)
+</pre></td>
+</tr>
+</table>
+
+## Example 4
+
+<table>
+<tr><th>GraphQL</th><th>Kotlin</th></tr>
+<tr>
+<td><pre lang="graphql">
+query TestOperation {
+  cat {
+    # Field type Cat is subtype of inline fragment
+    ... on Animal {
+      species
+    }
+  }
+}
+</pre></td>
+<td><pre lang="kotlin">
+public data class Data(public val cat: Cat?)
+data class Cat(val onAnimal: OnAnimal)
+</pre></td>
+</tr>
+</table>
+
+## Example 5
+
+<table>
+<tr><th>GraphQL</th><th>Kotlin</th></tr>
+<tr>
+<td><pre lang="graphql">
+query TestOperation {
+  animal {
+    ... on Cat {
+      species
+    }
+  }
+}
+</pre></td>
+<td><pre lang="kotlin">
+public data class Data(public val animal: IAnimal?)
+
+sealed interface IAnimal { val onCat: Animal.OnCat? }
+data class CatAnimal(val onCat: Animal.OnCat): IAnimal
+data class Animal(val onCat: Animal.OnCat?): IAnimal
+</pre></td>
+</tr>
+</table>
+
+
+
+## Example 6
+
+<table>
+<tr><th>GraphQL</th><th>Kotlin</th></tr>
+<tr>
+<td><pre lang="graphql">
+query TestOperation {
+  animal {
+    species
+    ... on WarmBlooded {
+      temperature
+    }
+    ... on Pet {
+      name
+    }
+  }
+}
+</pre></td>
+<td><pre lang="kotlin">
+public data class Data(public val animal: IAnimal?)
+
+// Abstraction Model
+sealed interface IAnimal { val onAnimal: Animal.IOnAnimal }
+// 3 TypeSet Models
+data class WarmBloodedAnimal(val onAnimal: Animal.IOnAnimal): IAnimal
+data class PetAnimal(val onAnimal: Animal.IOnAnimal): IAnimal
+data class WarmBloodedPetAnimal(val onAnimal: Animal.IOnAnimal): IAnimal
+// Fallback Model
+data class Animal(val onAnimal: IOnAnimal): IAnimal
+
+</pre></td>
+</tr>
+</table>
+
+
+## Example 7
+
+<table>
+<tr><th>GraphQL</th><th>Kotlin</th></tr>
+<tr>
+<td><pre lang="graphql">
+query TestOperation {
+  animal {
+    ... on Animal {
+      species
+      # Nested WarmBlooded
+      ... on WarmBlooded {
+        temperature
+      }
+      # Nested Pet
+      ... on Pet {
+        name
+      }
+    }
+  }
+}
+</pre></td>
+<td><pre lang="kotlin">
+
+// animal is not generated as polymorphic model because inline fragment ... on Animal is always assignable
+public data class Data(public val animal: Animal?)
+data class Animal(val onAnimal: IOnAnimal)
+
+// onAnimal field has two nested inline fragments -> generated as Polymorphic model
+public sealed interface IOnAnimal { 
+  val species: String,
+  val onWarmBlooded: OnAnimal.OnWarmBlooded?,
```

Comment:
I think there's no need to namespace `OnWarmBlooded` here as it will be accessible in the `OnAnimal` scope. `val onWarmBlooded: OnWarmBlooded?` should work (I included it in the suggestion below)